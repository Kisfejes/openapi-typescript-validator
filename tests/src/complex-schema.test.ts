import path from "path";
import fs from "fs";
const { generate } = require("openapi-typescript-validator");

describe("complex-schema", () => {
  const name = "complex";
  const generatedDir = path.join(__dirname, "../generated", name);
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    await generate({
      schemaFile: path.join(schemaDir, "complex-schema.json"),
      schemaType: "json",
      name,
      directory: generatedDir,
    });
  });

  test("schema should match", () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `${name}-schema.json`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });

  describe("decoders", () => {
    const decodersDir = path.join(generatedDir, `decoders`);

    test("file structure", () => {
      const dir = fs.readdirSync(decodersDir);
      expect(dir).toEqual([
        "ImageComponent",
        "Screen",
        "TitleComponent",
        "helpers.ts",
        "index.ts",
      ]);
    });

    test("index.ts", () => {
      const file = fs.readFileSync(path.join(decodersDir, `index.ts`), "utf8");
      expect(file).toMatchSnapshot();
    });
  });

  test("models should match", () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `${name}-models.ts`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });
});
