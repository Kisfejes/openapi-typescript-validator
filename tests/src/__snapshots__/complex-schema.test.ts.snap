// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex-schema decoders should match 1`] = `
"/* eslint-disable */
/* tslint-disable */
import { ErrorObject } from \\"ajv\\";
import * as types from \\"./complex-models\\";
import {
  ScreenValidator,
  TitleComponentValidator,
  ImageComponentValidator,
} from \\"./complex-validators\\";

// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// !!! AUTO GENERATED CODE, DON'T TOUCH !!!
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

interface Validator {
  (json: unknown): boolean;
  errors?: ErrorObject[];
}

function validateJson(
  json: any,
  validator: Validator,
  definitionName: string
): any {
  const jsonObject = typeof json === \\"string\\" ? JSON.parse(json) : json;

  if (validator(jsonObject)) {
    return jsonObject;
  }

  const jsonPreviewStr = (typeof json === \\"string\\"
    ? json
    : JSON.stringify(jsonObject)
  ).substring(0, 200);
  if (validator.errors) {
    throw Error(
      \`\${definitionName} \${errorsText(
        validator.errors
      )}. JSON: \${jsonPreviewStr}\`
    );
  }

  throw Error(
    \`\${definitionName} Unexpected data received. JSON: \${jsonPreviewStr}\`
  );
}

function errorsText(errors: ErrorObject[]): string {
  return errors
    .map((error) => \`\${error.instancePath}: \${error.message}\`)
    .join(\\"\\\\n\\");
}

// Decoders
export const ScreenDecoder = {
  definitionName: \\"Screen\\",
  schemaRef: \\"#/definitions/Screen\\",

  decode(json: any): types.Screen {
    return validateJson(
      json,
      ScreenValidator as Validator,
      ScreenDecoder.definitionName
    );
  },
};
export const TitleComponentDecoder = {
  definitionName: \\"TitleComponent\\",
  schemaRef: \\"#/definitions/TitleComponent\\",

  decode(json: any): types.TitleComponent {
    return validateJson(
      json,
      TitleComponentValidator as Validator,
      TitleComponentDecoder.definitionName
    );
  },
};
export const ImageComponentDecoder = {
  definitionName: \\"ImageComponent\\",
  schemaRef: \\"#/definitions/ImageComponent\\",

  decode(json: any): types.ImageComponent {
    return validateJson(
      json,
      ImageComponentValidator as Validator,
      ImageComponentDecoder.definitionName
    );
  },
};
"
`;

exports[`complex-schema models should match 1`] = `
"/* eslint-disable */

/* tslint:disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export type Component = TitleComponent | ImageComponent;

export interface Screen {
  components: Component[];
}
export interface TitleComponent {
  type: \\"title\\";
  title: string;
  subtitle?: string | null;
}
export interface ImageComponent {
  type: \\"image\\";
  url: string;
}
"
`;

exports[`complex-schema schema should match 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"title\\": \\"Schema\\",
  \\"definitions\\": {
    \\"Screen\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"components\\": {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"$ref\\": \\"#/definitions/Component\\"
          }
        }
      },
      \\"required\\": [
        \\"components\\"
      ],
      \\"$schema\\": \\"http://json-schema.org/draft-04/schema#\\"
    },
    \\"Component\\": {
      \\"anyOf\\": [
        {
          \\"$ref\\": \\"#/definitions/TitleComponent\\"
        },
        {
          \\"$ref\\": \\"#/definitions/ImageComponent\\"
        }
      ],
      \\"$schema\\": \\"http://json-schema.org/draft-04/schema#\\"
    },
    \\"TitleComponent\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"type\\": {
          \\"type\\": \\"string\\",
          \\"enum\\": [
            \\"title\\"
          ]
        },
        \\"title\\": {
          \\"type\\": \\"string\\"
        },
        \\"subtitle\\": {
          \\"type\\": [
            \\"string\\",
            \\"null\\"
          ]
        }
      },
      \\"required\\": [
        \\"type\\",
        \\"title\\"
      ],
      \\"$schema\\": \\"http://json-schema.org/draft-04/schema#\\"
    },
    \\"ImageComponent\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"type\\": {
          \\"type\\": \\"string\\",
          \\"enum\\": [
            \\"image\\"
          ]
        },
        \\"url\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"type\\",
        \\"url\\"
      ],
      \\"$schema\\": \\"http://json-schema.org/draft-04/schema#\\"
    }
  },
  \\"properties\\": {
    \\"Screen\\": {
      \\"$ref\\": \\"#/definitions/Screen\\"
    },
    \\"Component\\": {
      \\"$ref\\": \\"#/definitions/Component\\"
    },
    \\"TitleComponent\\": {
      \\"$ref\\": \\"#/definitions/TitleComponent\\"
    },
    \\"ImageComponent\\": {
      \\"$ref\\": \\"#/definitions/ImageComponent\\"
    }
  }
}"
`;

exports[`complex-schema validators.d.ts should match 1`] = `
"export function ScreenValidator(json: unknown): boolean;
export function TitleComponentValidator(json: unknown): boolean;
export function ImageComponentValidator(json: unknown): boolean;
"
`;

exports[`complex-schema validators.js should match 1`] = `
"
/* eslint-disable */

\\"use strict\\";export const ScreenValidator = validate16;const schema12 = {\\"type\\":\\"object\\",\\"properties\\":{\\"components\\":{\\"type\\":\\"array\\",\\"items\\":{\\"$ref\\":\\"#/definitions/Component\\"}}},\\"required\\":[\\"components\\"],\\"$schema\\":\\"http://json-schema.org/draft-04/schema#\\"};const schema13 = {\\"anyOf\\":[{\\"$ref\\":\\"#/definitions/TitleComponent\\"},{\\"$ref\\":\\"#/definitions/ImageComponent\\"}],\\"$schema\\":\\"http://json-schema.org/draft-04/schema#\\"};const schema14 = {\\"type\\":\\"object\\",\\"properties\\":{\\"type\\":{\\"type\\":\\"string\\",\\"enum\\":[\\"title\\"]},\\"title\\":{\\"type\\":\\"string\\"},\\"subtitle\\":{\\"type\\":[\\"string\\",\\"null\\"]}},\\"required\\":[\\"type\\",\\"title\\"],\\"$schema\\":\\"http://json-schema.org/draft-04/schema#\\"};const schema15 = {\\"type\\":\\"object\\",\\"properties\\":{\\"type\\":{\\"type\\":\\"string\\",\\"enum\\":[\\"image\\"]},\\"url\\":{\\"type\\":\\"string\\"}},\\"required\\":[\\"type\\",\\"url\\"],\\"$schema\\":\\"http://json-schema.org/draft-04/schema#\\"};const func0 = require(\\"ajv/dist/runtime/equal\\").default;function validate12(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;let valid0 = false;const _errs1 = errors;const _errs2 = errors;if(errors === _errs2){if(data && typeof data == \\"object\\" && !Array.isArray(data)){let missing0;if(((data.type === undefined) && (missing0 = \\"type\\")) || ((data.title === undefined) && (missing0 = \\"title\\"))){const err0 = {instancePath,schemaPath:\\"#/definitions/TitleComponent/required\\",keyword:\\"required\\",params:{missingProperty: missing0},message:\\"must have required property '\\"+missing0+\\"'\\"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}else {if(data.type !== undefined){let data0 = data.type;const _errs4 = errors;if(typeof data0 !== \\"string\\"){const err1 = {instancePath:instancePath+\\"/type\\",schemaPath:\\"#/definitions/TitleComponent/properties/type/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}if(!(data0 === \\"title\\")){const err2 = {instancePath:instancePath+\\"/type\\",schemaPath:\\"#/definitions/TitleComponent/properties/type/enum\\",keyword:\\"enum\\",params:{allowedValues: schema14.properties.type.enum},message:\\"must be equal to one of the allowed values\\"};if(vErrors === null){vErrors = [err2];}else {vErrors.push(err2);}errors++;}var valid2 = _errs4 === errors;}else {var valid2 = true;}if(valid2){if(data.title !== undefined){const _errs6 = errors;if(typeof data.title !== \\"string\\"){const err3 = {instancePath:instancePath+\\"/title\\",schemaPath:\\"#/definitions/TitleComponent/properties/title/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err3];}else {vErrors.push(err3);}errors++;}var valid2 = _errs6 === errors;}else {var valid2 = true;}if(valid2){if(data.subtitle !== undefined){let data2 = data.subtitle;const _errs8 = errors;if((typeof data2 !== \\"string\\") && (data2 !== null)){const err4 = {instancePath:instancePath+\\"/subtitle\\",schemaPath:\\"#/definitions/TitleComponent/properties/subtitle/type\\",keyword:\\"type\\",params:{type: schema14.properties.subtitle.type},message:\\"must be string,null\\"};if(vErrors === null){vErrors = [err4];}else {vErrors.push(err4);}errors++;}var valid2 = _errs8 === errors;}else {var valid2 = true;}}}}}else {const err5 = {instancePath,schemaPath:\\"#/definitions/TitleComponent/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"};if(vErrors === null){vErrors = [err5];}else {vErrors.push(err5);}errors++;}}var _valid0 = _errs1 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs10 = errors;const _errs11 = errors;if(errors === _errs11){if(data && typeof data == \\"object\\" && !Array.isArray(data)){let missing1;if(((data.type === undefined) && (missing1 = \\"type\\")) || ((data.url === undefined) && (missing1 = \\"url\\"))){const err6 = {instancePath,schemaPath:\\"#/definitions/ImageComponent/required\\",keyword:\\"required\\",params:{missingProperty: missing1},message:\\"must have required property '\\"+missing1+\\"'\\"};if(vErrors === null){vErrors = [err6];}else {vErrors.push(err6);}errors++;}else {if(data.type !== undefined){let data3 = data.type;const _errs13 = errors;if(typeof data3 !== \\"string\\"){const err7 = {instancePath:instancePath+\\"/type\\",schemaPath:\\"#/definitions/ImageComponent/properties/type/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err7];}else {vErrors.push(err7);}errors++;}if(!(data3 === \\"image\\")){const err8 = {instancePath:instancePath+\\"/type\\",schemaPath:\\"#/definitions/ImageComponent/properties/type/enum\\",keyword:\\"enum\\",params:{allowedValues: schema15.properties.type.enum},message:\\"must be equal to one of the allowed values\\"};if(vErrors === null){vErrors = [err8];}else {vErrors.push(err8);}errors++;}var valid4 = _errs13 === errors;}else {var valid4 = true;}if(valid4){if(data.url !== undefined){const _errs15 = errors;if(typeof data.url !== \\"string\\"){const err9 = {instancePath:instancePath+\\"/url\\",schemaPath:\\"#/definitions/ImageComponent/properties/url/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err9];}else {vErrors.push(err9);}errors++;}var valid4 = _errs15 === errors;}else {var valid4 = true;}}}}else {const err10 = {instancePath,schemaPath:\\"#/definitions/ImageComponent/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"};if(vErrors === null){vErrors = [err10];}else {vErrors.push(err10);}errors++;}}var _valid0 = _errs10 === errors;valid0 = valid0 || _valid0;}if(!valid0){const err11 = {instancePath,schemaPath:\\"#/anyOf\\",keyword:\\"anyOf\\",params:{},message:\\"must match a schema in anyOf\\"};if(vErrors === null){vErrors = [err11];}else {vErrors.push(err11);}errors++;validate12.errors = vErrors;return false;}else {errors = _errs0;if(vErrors !== null){if(_errs0){vErrors.length = _errs0;}else {vErrors = null;}}}validate12.errors = vErrors;return errors === 0;}function validate16(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == \\"object\\" && !Array.isArray(data)){let missing0;if((data.components === undefined) && (missing0 = \\"components\\")){validate16.errors = [{instancePath,schemaPath:\\"#/required\\",keyword:\\"required\\",params:{missingProperty: missing0},message:\\"must have required property '\\"+missing0+\\"'\\"}];return false;}else {if(data.components !== undefined){let data0 = data.components;const _errs1 = errors;if(errors === _errs1){if(Array.isArray(data0)){var valid1 = true;const len0 = data0.length;for(let i0=0; i0<len0; i0++){const _errs3 = errors;if(!(validate12(data0[i0], {instancePath:instancePath+\\"/components/\\" + i0,parentData:data0,parentDataProperty:i0,rootData}))){vErrors = vErrors === null ? validate12.errors : vErrors.concat(validate12.errors);errors = vErrors.length;}var valid1 = _errs3 === errors;if(!valid1){break;}}}else {validate16.errors = [{instancePath:instancePath+\\"/components\\",schemaPath:\\"#/properties/components/type\\",keyword:\\"type\\",params:{type: \\"array\\"},message:\\"must be array\\"}];return false;}}}}}else {validate16.errors = [{instancePath,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"}];return false;}}validate16.errors = vErrors;return errors === 0;}export const TitleComponentValidator = validate18;function validate18(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == \\"object\\" && !Array.isArray(data)){let missing0;if(((data.type === undefined) && (missing0 = \\"type\\")) || ((data.title === undefined) && (missing0 = \\"title\\"))){validate18.errors = [{instancePath,schemaPath:\\"#/required\\",keyword:\\"required\\",params:{missingProperty: missing0},message:\\"must have required property '\\"+missing0+\\"'\\"}];return false;}else {if(data.type !== undefined){let data0 = data.type;const _errs1 = errors;if(typeof data0 !== \\"string\\"){validate18.errors = [{instancePath:instancePath+\\"/type\\",schemaPath:\\"#/properties/type/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"}];return false;}if(!(data0 === \\"title\\")){validate18.errors = [{instancePath:instancePath+\\"/type\\",schemaPath:\\"#/properties/type/enum\\",keyword:\\"enum\\",params:{allowedValues: schema14.properties.type.enum},message:\\"must be equal to one of the allowed values\\"}];return false;}var valid0 = _errs1 === errors;}else {var valid0 = true;}if(valid0){if(data.title !== undefined){const _errs3 = errors;if(typeof data.title !== \\"string\\"){validate18.errors = [{instancePath:instancePath+\\"/title\\",schemaPath:\\"#/properties/title/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"}];return false;}var valid0 = _errs3 === errors;}else {var valid0 = true;}if(valid0){if(data.subtitle !== undefined){let data2 = data.subtitle;const _errs5 = errors;if((typeof data2 !== \\"string\\") && (data2 !== null)){validate18.errors = [{instancePath:instancePath+\\"/subtitle\\",schemaPath:\\"#/properties/subtitle/type\\",keyword:\\"type\\",params:{type: schema14.properties.subtitle.type},message:\\"must be string,null\\"}];return false;}var valid0 = _errs5 === errors;}else {var valid0 = true;}}}}}else {validate18.errors = [{instancePath,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"}];return false;}}validate18.errors = vErrors;return errors === 0;}export const ImageComponentValidator = validate19;function validate19(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == \\"object\\" && !Array.isArray(data)){let missing0;if(((data.type === undefined) && (missing0 = \\"type\\")) || ((data.url === undefined) && (missing0 = \\"url\\"))){validate19.errors = [{instancePath,schemaPath:\\"#/required\\",keyword:\\"required\\",params:{missingProperty: missing0},message:\\"must have required property '\\"+missing0+\\"'\\"}];return false;}else {if(data.type !== undefined){let data0 = data.type;const _errs1 = errors;if(typeof data0 !== \\"string\\"){validate19.errors = [{instancePath:instancePath+\\"/type\\",schemaPath:\\"#/properties/type/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"}];return false;}if(!(data0 === \\"image\\")){validate19.errors = [{instancePath:instancePath+\\"/type\\",schemaPath:\\"#/properties/type/enum\\",keyword:\\"enum\\",params:{allowedValues: schema15.properties.type.enum},message:\\"must be equal to one of the allowed values\\"}];return false;}var valid0 = _errs1 === errors;}else {var valid0 = true;}if(valid0){if(data.url !== undefined){const _errs3 = errors;if(typeof data.url !== \\"string\\"){validate19.errors = [{instancePath:instancePath+\\"/url\\",schemaPath:\\"#/properties/url/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"}];return false;}var valid0 = _errs3 === errors;}else {var valid0 = true;}}}}else {validate19.errors = [{instancePath,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"}];return false;}}validate19.errors = vErrors;return errors === 0;}
"
`;
